[AMORO-1123] Replace fastjson with jackson

klion26 opened the pull request:
<!--
Thanks for sending a pull request!

Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://amoro.netease.com/how-to-contribute/
  2. If the PR is related to an issue in https://github.com/NetEase/amoro/issues, add '[AMORO-XXXX]' in your PR title, e.g., '[AMORO-XXXX] Your PR title ...'.
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][AMORO-XXXX] Your PR title ...'.
-->

## Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you add a feature, you can talk about its use case.
  2. If you fix a bug, you can clarify why it is a bug.
  3. Use Fix/Resolve/Close #{ISSUE_NUMBER} to link this PR to its related issue
-->

Close #1123 .

## Brief change log
<!--
Clearly describe the changes made in modules, classes, methods, etc.
-->

- Replace fastjson with jackson

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (yes / no)
- If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

klion26 commented:
This is a draft version, mainly for compare

klion26 commented on pom.xml:
```diff
@@ -104,6 +104,7 @@
         <curator.version>5.6.0</curator.version>
         <zookeeper.version>3.9.1</zookeeper.version>
         <fastjson.version>1.2.83</fastjson.version>
+        <jackson.version>2.13.5</jackson.version>
```
The version picked from Spark 3, can change this if we needed

zhoujinsong replied:
It seems that different projects may depend on different versions of Jackson, just like Guava. Should we do better shade them for Amoro and use the shaded classes?

zhoujinsong replied:
Just like what Paimon had do: https://mvnrepository.com/artifact/org.apache.paimon/paimon-shade-jackson-2

klion26 commented:
cc @XBaith @zhongqishang @zhoujinsong

shidayang commented on ams/server/src/test/java/com/netease/arctic/server/util/JacksonUtilsTest.java:
```diff
@@ -0,0 +1,165 @@
+    assertTrue(JacksonUtils.getBoolean(jsonNode, "NboolValue", true));
+
+    Map<String, String> yamlMap =
+        new Yaml().loadAs(this.getClass().getResourceAsStream("/test.yaml"), Map.class);
```
Can we use a String directly instead of a YAML file?

zhoujinsong commented on ams/api/src/main/java/com/netease/arctic/ams/api/client/ArcticThriftUrl.java:
```diff
@@ -45,6 +48,14 @@ public class ArcticThriftUrl {
   private static final Logger logger = LoggerFactory.getLogger(ArcticThriftUrl.class);
   private static final Pattern PATTERN = Pattern.compile("zookeeper://(\\S+)/([\\w-]+)");
+
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
```
Can we move the `JacksonUtils` implementation into the `ams-api` module? So we can use it directly here.

klion26 replied:
Seems we use JacksonUtils more in `ams-server` module, and only one usage in `ams-api`, and `ams-server` does not dependent on `ams-api`, so didn't create it in `ams-api`
I'll try to add dependent `ams-api` in `ams-server` and move `JacksonUtls`

zhoujinsong commented on core/pom.xml:
```diff
@@ -39,6 +39,11 @@
 
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-core</artifactId>
```
Do we still need to declare the dependency here if the module `core` had dependent on `ams-api`.

klion26 replied:
IIUC, we need to explict define the denpendency here, otherwise we can't find it when compiling. will double check here after has shaded all the jackson involded here

zhoujinsong commented on mixed/spark/v3.2/spark/pom.xml:
```diff
@@ -101,6 +101,14 @@
         </dependency>
 
+        <dependency>
+            <!-- Scala module 2.12.3 requires Jackson Databind version >= 2.12.0 and < 2.13.0 -->
```
It seems to be that the scala version here is 2.12.15.

klion26 replied:
Spark 3.2 means using 2.12.3, so it requires Jackson version [2.12.0, 2.13.0) , the comment here want to record why do we need to add a provided dependency here

zhoujinsong commented on mixed/spark/v3.3/spark-runtime/pom.xml:
```diff
@@ -63,7 +63,7 @@
                                     <include>com.netease.amoro:amoro-ams-api</include>
                                     <include>com.netease.amoro:amoro-mixed-hive</include>
-                                    <include>com.alibaba:fastjson</include>
+                                    <include>com.fasterxml.jackson.core:jackson-databind</include>
```
Why does Spark 3.3 require relocation of this class, while Spark 3.2 does not?

klion26 replied:
maybe after shade the jackson we involed, we can remove this here.

IIUC, this added because of failed test

zhoujinsong reviewed (commented):
@klion26 Thanks a lot for your contribution!

I left some comments, please take another look.

zhoujinsong commented on ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/LoginController.java:
```diff
@@ -18,7 +18,7 @@
 package com.netease.arctic.server.dashboard.controller;
 
-import com.alibaba.fastjson.JSONObject;
+import com.fasterxml.jackson.databind.JsonNode;
```
Do we need to import the shaded class here?

klion26 replied:
yes, we should use the shaded class

zhoujinsong commented on core/src/test/java/com/netease/arctic/utils/JacksonUtilsTest.java:
```diff
@@ -0,0 +1,196 @@
+
+package com.netease.arctic.utils;
+
+import static org.junit.Assert.*;
```
We should try to avoid importing the entire package whenever possible.

klion26 replied:
my bad, changed my IDEA's config to test the check for wildcard import, correct now.

zhoujinsong reviewed (commented):
@klion26 Thanks a lot for your contribution!
The code is overall fine, but I still have a few minor suggestions. Please take a look.

zhoujinsong commented:
After this PR, perhaps we can collaborate on optimizing the current check style rules, such as：

- check whether the entire package is imported
- check whether the shaded classes are imported rather than the original class for guvava and jackson.

zhongqishang commented on ams/server/src/main/java/com/netease/arctic/server/ArcticServiceContainer.java:
```diff
@@ -496,20 +502,22 @@ private void setIcebergSystemProperties() {
     @SuppressWarnings("unchecked")
```
We can delete this `SuppressWarnings` together.

zhongqishang commented on core/src/main/java/com/netease/arctic/utils/JacksonUtils.java:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+
+  /** Deserialize a json object from string which serialized using {@link toJSONString()}. */
```
Unable link.

zhongqishang reviewed (commented):
Thanks for your contribution, I left some comments.

zhongqishang commented:
@klion26 I also find some shade plugin include the fastjson in pom.xml, we should modify those config together.

zhongqishang commented on core/src/test/java/com/netease/arctic/utils/JacksonUtilsTest.java:
```diff
@@ -0,0 +1,196 @@
+
+    Map<String, Object> yamlMap = new HashMap<>();
+    Map<String, Object> amsNode = new HashMap<>();
+    amsNode.put("arctic.ams.server-host.prefix", "127.");
```
We shouldn't use the old name `arctic`, this will be import additional renaming work later.

klion26 commented on ams/server/src/main/java/com/netease/arctic/server/dashboard/controller/LoginController.java:
```diff
@@ -46,7 +46,7 @@ public void getCurrent(Context ctx) {
-    JSONObject postBody = ctx.bodyAsClass(JSONObject.class);
+    JsonNode postBody = ctx.bodyAsClass(JsonNode.class);
     if (adminUser.equals(postBody.get("user"))
         && (adminPassword.equals(postBody.get("password")))) {
```
changed for now, thanks for catch this.

klion26 commented:
@zhoujinsong @zhongqishang thanks for the review, addressed the comments, please have another look when you're free.

@zhoujinsong will add checksytle(wildcard import/shaded import etc)after this pr

zhongqishang commented:
We should add paimon jackson shade jar to spark runtime include? @klion26 @zhoujinsong

klion26 commented:
@zhongqishang IIUC, currently spark contains jaskson itself, so we don't need to add the shaded jackson into spark runtime.

zhoujinsong commented:
> Should we add paimon jackson shade jar to spark runtime include ? @klion26 @zhoujinsong

Thanks a lot for pointing this out. @zhongqishang 
I am afraid we should add the paimon shaded jackson artifact into the shading rules in Spark and Flink runtime modules like we did for iceberg shaded guava artifact.
The reason is that the engine modules dependent on `ams-core` and may need to parse json for connecting to ams server.

zhoujinsong commented on core/src/main/java/com/netease/arctic/utils/JacksonUtils.java:
```diff
@@ -0,0 +1,140 @@
+import java.util.Map;
+
+/** Util class for Jackson. */
+public class JacksonUtils {
```
Can we rename the class name to `JacksonUtil` to keep consistent with other util classes in the project.

zhoujinsong reviewed (commented):
Another small suggestion, PTAL. @klion26

zhongqishang reviewed (approved):
LGTM.

klion26 commented:
@zhoujinsong @zhongqishang @XBaith @shidayang thanks all for the detailed review and merging!
